### PR TITLE
membuf: add SliceLocation to reduce GC pressure

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/48779

Problem Summary:

### What is changed and how it works?

move external package's kvLocation to membuf. And re-evaluate the tradeoff of this design. My conclusion is

- storing `SliceLocation` will reduce memory consumption: 2.34GB -> 1.17GB for 100M objects, where object is 10 byte 
- `SliceLocation` will make sorting slower: 713 ms -> 953 ms for 3M objects

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

## benchmark about storing in memory

```
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/br/pkg/membuf
cpu: Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz
BenchmarkStoreSlice
BenchmarkStoreSlice-12       	       1	1505804316 ns/op
BenchmarkStoreLocation
BenchmarkStoreLocation-12    	       2	 621385270 ns/op
PASS
```

CPU profile:

store slice:

image

store location:

image
runtime.memclrNoHeapPointers is amortized to alloc's invocation so scan objects are negligible

Memory profile:

store slice: 2.34GB
store location: 1.17GB

## benchmark about storing and sorting

```bash
$ for name in BenchmarkSortSliceWithGC BenchmarkSortLocationWithGC BenchmarkSortLocationPointerWithGC; do
go test ./br/pkg/membuf -bench $name -benchtime 1x -cpuprofile /tmp/$name.prof
done
```

```
goos: darwin
goarch: arm64
pkg: github.com/pingcap/tidb/br/pkg/membuf
BenchmarkSortSliceWithGC-8   	       1	 713820667 ns/op
PASS
ok  	github.com/pingcap/tidb/br/pkg/membuf	1.356s
goos: darwin
goarch: arm64
pkg: github.com/pingcap/tidb/br/pkg/membuf
BenchmarkSortLocationWithGC-8   	       1	 953731750 ns/op
PASS
ok  	github.com/pingcap/tidb/br/pkg/membuf	1.214s
goos: darwin
goarch: arm64
pkg: github.com/pingcap/tidb/br/pkg/membuf
BenchmarkSortLocationPointerWithGC-8   	       1	1272493792 ns/op
PASS
ok  	github.com/pingcap/tidb/br/pkg/membuf	1.515s
```

[bench-profiles.zip](https://github.com/pingcap/tidb/files/13432819/bench-profiles.zip)

We can see the duration in GC (O(N) where N is number of objects) is quite less than sorting (O(NlogN) where N is number of objects). Introducing `SliceLocation` will increase the duration of swapping and getting sort key.

BenchmarkSortSliceWithGC:
image

BenchmarkSortLocationWithGC:
image

Side effects
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/membuf/buffer.go [hunk 1]
```diff
@@ -130,64 +125,130 @@ func (p *Pool) TotalSize() int64 {
...
 	}
+}
 
-	b.curBufLen = len(b.curBuf)
-	b.curIdx = 0
+// NewBuffer creates a new buffer in current pool.
+func (p *Pool) NewBuffer(opts ...BufferOption) *Buffer {
+	b := &Buffer{
+		pool:          p,
+		blocks:        make([][]byte, 0, 128),
```
Comment (tangenta) [inline br/pkg/membuf/buffer.go:155]: Tiny improvement: maybe move this initialization after all `BufferOption` are applied? Thus we don't waste this slice if `WithMemoryLimit` exists.

**Review by tangenta — APPROVED** [review]

**Comment by lance6716** [comment]
/hold

new benchmark result is not expected

**Review by D3Hunter — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**